smack-integration-test/src/main/java/org/jivesoftware/smack/roster/RosterIntegrationTest.java

```diff
@@ -104,7 +122,372 @@ private void checkIfAddedAndSubscribed(Collection<Jid> addresses) {
+        "to a value of \"both\", as explained under Appendix A.)")
+    public void testPresenceApprovalYieldsRosterPush2() throws Exception {
+        IntegrationTestRosterUtil.ensureBothAccountsAreNotInEachOthersRoster(conOne, conTwo);
+
+        // Setup fixture: establish one-way subscription.
+        rosterOne.setSubscriptionMode(Roster.SubscriptionMode.accept_all);
+
+        final SimpleResultSyncPoint fixtureComplete = new SimpleResultSyncPoint();
+        RosterListener rosterListenerTwo = new AbstractRosterListener() {
+            @Override
+            public void entriesAdded(Collection<Jid> addresses) {
+                checkIfAdded(addresses);
+            }
+            @Override
+            public void entriesUpdated(Collection<Jid> addresses) {
+                checkIfAdded(addresses);
+            }
+            private void checkIfAdded(Collection<Jid> addresses) {
+                for (Jid jid : addresses) {
+                    final BareJid bareJid = conOne.getUser().asBareJid();
+                    if (!jid.equals(bareJid)) {
+                        continue;
+                    }
+                    if (rosterTwo.getEntry(bareJid) == null) {
+                        continue;
+                    }
+                    if (rosterTwo.getEntry(bareJid).getType() == ItemType.none) {
+                        continue;
+                    }
+                    fixtureComplete.signal();
+                    rosterTwo.removeRosterListener(this);
+                }
+            }
+        };
+        rosterTwo.addRosterListener(rosterListenerTwo);
+
+        final Presence subscribeOne = PresenceBuilder.buildPresence()
+                .ofType(Presence.Type.subscribe)
+                .to(conOne.getUser().asBareJid())
+                .build();
+        try {
+            conTwo.sendStanza(subscribeOne);
+
+            fixtureComplete.waitForResult(2 * connection.getReplyTimeout());
+        } finally {
+            rosterOne.setSubscriptionMode(Roster.getDefaultSubscriptionMode());
+            rosterTwo.removeRosterListener(rosterListenerTwo);
+        }
+
+        // Setup fixture is now complete. Execute the test.
+        rosterTwo.setSubscriptionMode(Roster.SubscriptionMode.accept_all);
+
+        final ResultSyncPoint<RosterEntry, Exception> updated = new ResultSyncPoint<>();
+
+        rosterListenerTwo = new AbstractRosterListener() {
+            @Override
+            public void entriesUpdated(Collection<Jid> addresses) {
+                for (Jid jid : addresses) {
+                    if (!jid.equals(conOne.getUser().asBareJid())) {
+                        continue;
+                    }
+                    BareJid bareJid = conOne.getUser().asBareJid();
+                    updated.signal(rosterTwo.getEntry(bareJid));
+                }
+            }
+        };
+        rosterTwo.addRosterListener(rosterListenerTwo);
+
+        final Presence subscribeTwo = PresenceBuilder.buildPresence()
```

Comment:
Please use conTwo.getStanzaFactory() to build XMPP stanzas send over a particular connection.